File: examples/game-server-demo-1/GameServerDemo1/src/main/java/com/huya/ig/game/logic/Room.java
package com.huya.ig.game.logic;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.huya.ig.common.huyaapi.MsgListener;
import com.huya.ig.common.huyaapi.MsgSubscriber;
import com.huya.ig.common.network.util.ChannelUtil;
import com.huya.ig.common.utils.CommonComponent;
import io.netty.util.internal.ConcurrentSet;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;

import java.net.URI;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/**
 * 游戏房间，一般一个直播间对应一个游戏房间，同一个直播间游戏的玩家都会加入这个游戏房间
 * 核心的游戏玩法逻辑都在这里实现
 * @date 2020/5/21
 * @author wangpeng1@huya.com
 * @description 游戏房间，处理游戏逻辑
 *
 */
@Data
@Slf4j
public class Room {

    // 房间号，就是主播直播间房间号
    private String roomId;
    // 主播unionId
    private String presenterUid;
    // 游戏开始时间
    private long gameStartedAt;
    private ERoomState state;
    private Set<String> allUids = new ConcurrentSet<>();
    private Map<String, Player> signedPlayers = new ConcurrentHashMap<>();
    private Gson gson = new Gson();
    private MsgSubscriber subscriber;
    private int reconnTimes = 0;

    /**
     * 玩家连接房间
     * @param uid 玩家unionId
     */
    public void onJoin(String uid){
        allUids.add(uid);
        Map<String, Object> data = new HashMap<>();
        data.put("uid", uid);
        data.put("gaming", state == ERoomState.Gaming);
        data.put("signedUp", signedPlayers.containsKey(uid));
        broadcast(Protocol.S2CPlayerJoin.uri, gson.toJson(data));
    }

    /**
     * 主播开始游戏
     * @param uid
     */
    public void start(String uid){
        if(uid.equals(presenterUid) && state != ERoomState.Gaming && signedPlayers.size() >= GameConfig.minPlayerCnt){
            state = ERoomState.Gaming;
            gameStartedAt = System.currentTimeMillis();
            broadcast(Protocol.S2CGameStart.uri, "");

            // 启动游戏结束定时器
            RoomManager.getInstance().getSchduler().schedule(new Runnable() {
                @Override
                public void run() {
                    gameOver();
                }
            }, GameConfig.gameDuration, TimeUnit.SECONDS);
        }
    }

    /**
     * 玩家报名参加游戏
     * @param uid
     */
    public void signup(String uid, String payload){
        boolean success = true;
        if(signedPlayers.size() >= GameConfig.maxPlayerCnt){
            success = false;
        }else {
            JsonObject data = gson.fromJson(payload, JsonObject.class);
            signedPlayers.putIfAbsent(uid, new Player(uid, 0,
                    data.get("nick").getAsString(),
                    data.get("avatar").getAsString()));
        }
        Map<String, Object> data = new HashMap<>();
        if(success){
            data.put("success", true);
            data.put("player", signedPlayers.get(uid));
            broadcast(Protocol.S2CPlayerSignup.uri, gson.toJson(data));
        }else{
            data.put("success", false);
            unicast(uid, Protocol.S2CPlayerSignup.uri, gson.toJson(data));
        }
    }

    public void playerScore(String uid, String payload){
        if(signedPlayers.containsKey(uid) && state == ERoomState.Gaming){
            JsonObject data = gson.fromJson(payload, JsonObject.class);
            signedPlayers.get(uid).setScore(data.get("score").getAsInt());

            broadcast(Protocol.S2CRealtimeRank.uri, gson.toJson(rankList()));
        }
    }

    private List<Player> rankList(){
        return signedPlayers.values().stream().sorted(new Comparator<Player>() {
            @Override
            public int compare(Player o1, Player o2) {
                return o2.getScore() - o1.getScore();
            }
        }).collect(Collectors.toList());
    }

    /**
     * 主播主动结束游戏
     * @param uid
     */
    public void gameover(String uid){
        if(uid.equals(presenterUid) && state == ERoomState.Gaming){
            gameOver();
        }
    }

    /**
     * 游戏倒计时结束
     */
    private void gameOver(){
        state = ERoomState.GameOver;
        if(subscriber != null){
            subscriber.disconnect();
        }
        RoomManager.getInstance().removeRoom(roomId);
        // 计算排行榜
        broadcast(Protocol.S2CGameOver.uri, gson.toJson(rankList()));
    }

    /**
     * 订阅直播间弹幕消息（如果有需要，在房间初始化时调用此方法即可）
     * 已经封装好订阅虎牙弹幕消息的网络模块
     * 业务上只需要处理onMsg回调即可。
     */
    private void subHuyaMsg(){
        try {
            subscriber = new MsgSubscriber(new URI(CommonComponent.getInstance().genHuyaWsUrl(roomId)),
                    new String[]{"getMessageNotice", "getSendItemNotice"}, new MsgListener() {
                @Override
                public void onMsg(String type, String data) {
                    try {
                        if (type.equals("getSendItemNotice")) {
                            JsonObject json = gson.fromJson(data, JsonObject.class);
                            int giftId = json.get("itemId").getAsInt();
                            int giftCnt = json.get("sendItemCount").getAsInt();
                            int giftValue = json.get("totalPay").getAsInt();
                            String unionId = json.get("unionId").getAsString();
                            log.info("虎牙websocket uid:{} 收到送礼消息:{}", unionId, data);
                            // todo 业务处理逻辑
                            // 比如送礼物才能参与报名游戏，那么这里收到送礼消息后，判断是指定礼物，执行join逻辑即可
                            // onJoin(unionId);
                        } else if (type.equals("getMessageNotice")) {
                            // https://dev.huya.com/docs#/%E5%BC%80%E6%94%BEAPI%E5%8D%8F%E8%AE%AE%E8%AF%B4%E6%98%8E?id=_4%e5%bc%b9%e5%b9%95%e4%ba%8b%e4%bb%b6%e5%af%b9%e5%ba%94%e9%80%9a%e7%94%a8%e5%9b%9e%e5%8c%85%e7%bb%93%e6%9e%84data%e5%ad%97%e6%ae%b5%ef%bc%8c%e5%af%b9%e5%ba%94notice%ef%bc%9agetmessagenotice
                            JsonObject json = gson.fromJson(data, JsonObject.class);
                            String unionId = json.get("unionId").getAsString();
                            log.info("虎牙websocket uid:{} 收到弹幕消息:{}", unionId, data);
                            // todo 业务处理逻辑
                            // 比如发送指定弹幕消息，才能触发某个业务逻辑，那么判断满足消息条件后，执行对应的函数即可
                            // onBizMsg(unionId, msg);
                        }
                    }catch (Exception e){
                        log.error("error@deal huya-danmu", e);
                    }
                }

                @Override
                public void onDisconnect() {
                    if(state != ERoomState.GameOver){
                        // 重连
                        RoomManager.getInstance().getSchduler().schedule(new Runnable() {
                            @Override
                            public void run() {
                                if(state == ERoomState.GameOver){
                                    return;
                                }
                                log.info("虎牙websocket断线重连 roomId:{}", roomId);
                                subHuyaMsg();
                                reconnTimes++;
                                if(reconnTimes >= 10){
                                    reconnTimes = 0;
                                }
                            }
                        }, 2+reconnTimes*2, TimeUnit.SECONDS);
                    }
                }
            }, roomId);
            subscriber.connect();
        } catch (Exception e) {
            log.error("error@connect huya open ws-api", e);
        }
    }

    /**
     * 房间消息广播
     * @param uri 协议号
     * @param data 数据
     */
    private void broadcast(int uri, String data){
        broadcast(uri, data, null);
    }

    /**
     * 房间消息广播
     * @param uri 协议号
     * @param data 数据
     * @param exUids 排除的玩家unionId列表
     */
    private void broadcast(int uri, String data, Set<String> exUids){
        ChannelUtil.broadcast(roomId, uri, data, allUids, exUids);
    }

    /**
     * 按玩家unionId单播
     * @param uid 玩家unionId
     * @param uri 协议号
     * @param data 数据
     */
    private void unicast(String uid, int uri, String data){
        ChannelUtil.unicast(roomId, uri, data, uid);
    }

    enum ERoomState {
        Gaming, GameOver
    }

}

File: examples/game-common-demo-a/server-python/src/logic/room.py
import threading
import json
import time
import asyncio

import base64

from enum import Enum
from config import DefaultConfig
from logger import log
from .protocol import PROTOCOL
from util import ws_channel_util
from util import jwt_util
from hy_subscriber.subscriber import LiveSubscriber
from .player import Player, PlayerEncoder


class ERoomState(Enum):
    """ 房间状态枚举定义 """
    Gaming = 1
    GameOver = 2


class Room(object):
    """ 
    游戏房间，一般一个直播间对应一个游戏房间，同一个直播间游戏的玩家都会加入这个游戏房间
    核心的游戏玩法逻辑都在这里实现
    """

    # 房间号，就是主播直播间房间号
    roomId = None
    # 主播unionId
    presenterUid = None
    # 游戏开始时间
    gameStartedAt = None
    # 房间当前状态
    state = None
    # 所有连接到房间的观众和主播id列表
    allUids = []
    # 当前参与对局的观众和主播Player列表
    signedPlayers = {}
    #
    reconnTimes = 0
    # 游戏计时检测定时器
    timer = None

    # 监听虎牙直播间送礼弹幕等事件
    liveSubscriber = None

    # 重连虎牙直播间订阅服务的次数
    reconnTimes = 0

    def __init__(self, roomId, presenterUid):
        self.roomId = roomId
        self.presenterUid = presenterUid
        self.signedPlayers ={}
        self.allUids = []
        self.reconnTimes = 0
        self.state = None
        self.gameStartedAt =None
        self.liveSubscriber = None
        self.timer = None
        self.__connect_live_subscriber()

    def __connect_live_subscriber(self):
        """ 连虎牙直播间订阅服务 """
        if DefaultConfig.LIVE_SUBSCRIBER:
            wsUrl = jwt_util.build_huya_wsUrl(self.roomId)
            self.liveSubscriber = LiveSubscriber(
                self.roomId, wsUrl, DefaultConfig.LIVE_SUBSCRIBER_TOPICS,
                self.__live_message, self.__live_disconnect)
            self.liveSubscriber.connect()

    async def onJoin(self, uid):
        """ 玩家连接房间 """
        if not self.allUids.__contains__(uid):
            self.allUids.append(uid)
        data = {}
        data['uid'] = uid
        data['gaming'] = (self.state == ERoomState.Gaming)
        data['signedUp'] = (self.signedPlayers.__contains__(uid))

        players = []
        if self.signedPlayers:
            for i in self.signedPlayers:
                players.append(self.signedPlayers[i])

        data['players'] = players
        # 房间内广播，玩家加入
        await self.__broadcast(PROTOCOL.S2CPlayerJoin, json.dumps(data, cls=PlayerEncoder))

    async def start(self, uid):
        """ 开始游戏 """
        if uid == self.presenterUid and self.state != ERoomState.Gaming and len(self.signedPlayers) >= DefaultConfig.MIN_PLAYER_CNT:
            self.state = ERoomState.Gaming
            self.gameStartedAt = int(time.time()*1000)  # 当前毫秒值
            # 房间内广播，游戏开始
            await self.__broadcast(PROTOCOL.S2CGameStart)

        # 启动计时检测
        self.timer = threading.Timer(
            DefaultConfig.GAME_DURATION, self.__game_over_task)
        self.timer.start()


    def __game_over_task(self):
        """ 创建异步任务 """
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        try:
            loop.run_until_complete(self.__gameOver())
        finally:
            loop.run_until_complete(loop.shutdown_asyncgens())
            loop.close()

    async def signup(self, uid, data):
        """ 加入游戏 """
        canSignup = True
        # 未加入游戏，且游戏人数没有满
        if not self.signedPlayers.__contains__(uid) and len(self.signedPlayers) >= DefaultConfig.MAX_PLAYER_CNT:
            canSignup = False

        if canSignup:
            # 加入列表
            if not self.signedPlayers.__contains__(uid):
                player = Player(uid)
                nick = data['nick']
                if nick:
                    nick = base64.b64decode(nick).decode("utf-8")
                player.nick = nick
                player.avatar = data['avatar']
                self.signedPlayers[uid] = player
            data = {}
            data['success'] = True
            data['player'] = self.signedPlayers[uid]
            # 广播加入游戏成功
            await self.__broadcast(PROTOCOL.S2CPlayerSignup, json.dumps(data, cls=PlayerEncoder))
        else:
            data = {}
            data['success'] = False
            # 发给当前用户，加入失败
            await self.__unicast(uid, PROTOCOL.S2CPlayerSignup, json.dumps(data, cls=PlayerEncoder))

    async def __gameOver(self):
        """ 游戏结束 """
        log.info('game timeout,gameOver,{}'.format(self.roomId))

        # 广播游戏结束，可附带排行榜等数据
        await self.__broadcast(PROTOCOL.S2CGameOver,  json.dumps(self.__build_rank(),cls=PlayerEncoder) )

        self.state = ERoomState.GameOver

        if DefaultConfig.LIVE_SUBSCRIBER:
            self.liveSubscriber.disconnect()

        from logic.room_mgr import room_manager
        room_manager.remove_room(self.roomId)

    async def gameOver(self, uid):
        """ 主播主动结束游戏 """
        if uid == self.presenterUid and self.state == ERoomState.Gaming:
            await self.__gameOver()

    async def playerScore(self, uid, data):
        """ 玩家上报得分 """
        if self.signedPlayers.__contains__(uid) and self.state == ERoomState.Gaming:
            self.signedPlayers[uid].score = int(data['score'])
            data = self.__build_rank()
            await self.__broadcast(PROTOCOL.S2CRealtimeRank, json.dumps(data,cls=PlayerEncoder))

    def __build_rank(self):
        """ 构造实时得分排行榜 """

        players = []
        if self.signedPlayers:
            for i in self.signedPlayers:
                players.append(self.signedPlayers[i])
            
        data = sorted(players,  key=lambda player: player.score, reverse=True)
        return data

    async def __broadcast(self, uri,  data=None,  exUids=None):
        """ 广播 """
        await ws_channel_util.broadcast(self.roomId, uri, data, self.allUids)

    #
    async def __unicast(self, uid, uri,  data=None):
        """ 发送给指定用户 """
        await ws_channel_util.unicast(self.roomId, uid, uri, data)

    def __live_message(self, notice, data):
        """ 收到虎牙直播间订阅消息回调 """

        log.info("收到直播间订阅消息:{},{},{}".format(self.roomId, notice, str(data)))
        if notice == "getSendItemNotice":
            # data dict 结构：
            # 'badgeName':'叶妃'
            # 'fansLevel':1
            # 'itemId':4
            # 'itemName':'虎粮'
            # 'nobleLevel':0
            # 'presenterNick':'xxx'
            # 'roomId':15687938
            # 'sendItemComboHits':1
            # 'sendItemCount':1
            # 'sendNick':'xxx'
            # 'senderAvatarurl':'http://huyaimg.msstatic.com/avatar/1021/...1509880585'
            # 'senderLevel':2
            # 'totalPay':0
            # 'unionId':'unck5...Bo'

            unionId = data['unionId']  # 发送者unionId
            sendNick = data['sendNick']  # 发送者昵称
            senderAvatarurl = data['senderAvatarurl']  # 发送者头像
            fansLevel = data['fansLevel']  # 粉丝等级
            itemId = data['itemId']  # 礼物id
            itemName = data['itemName']  # 礼物id
            sendItemComboHits = data['sendItemComboHits']  # 送礼连击数
            sendItemCount = data['sendItemCount']  # 送礼个数
            totalPay = data['totalPay']  # 礼物价值

            log.info("监听到直播间送礼消息:用户id：{}，昵称：{}，礼物Id:{}，礼物名称:{}，礼物数:{}，连击数:{}，totalPay:{}".format(
                unionId, sendNick, itemId, itemName, sendItemCount, sendItemComboHits, totalPay))

            # @TODO 业务逻辑。。。

        elif notice == "getMessageNotice":
            # 收到 弹幕消息
            # 'data':{'badgeName': '', 'content': 'xxx', 'fansLevel': 0, 'nobleLevel': 0, 'roomId': 12345678, 'sendNick': 'xxx', 'senderAvatarUrl': 'http://huyaimg.msst...521768567', 'senderGender': 0, 'senderLevel': 1, 'showMode': 0, 'unionId': 'unrm3gdloD/ZmcQCXRH...H/zi0j48p'}
            unionId = data['unionId']  # 发送者unionId
            sendNick = data['sendNick']  # 发送者昵称
            senderAvatarUrl = data['senderAvatarUrl']  # 发送者头像
            senderGender = data['senderGender']  # 发送者性别
            content = data['content']  # 弹幕内容
            log.info("监听到直播间弹幕消息:用户id：{}，昵称：{}，弹幕内容:{}".format(
                unionId, sendNick, content))

            # @TODO 业务逻辑。。。

        else:
            log.info("监听到直播间{}消息,内容:{}".format(notice, str(data)))

    def __live_disconnect(self):
        """ 订阅服务连接断开回调 """

        log.warn("监听直播间订阅消息连接断开:{}".format(self.roomId))
        if self.state != ERoomState.GameOver:
            reTrytimer = threading.Timer(
                2+self.reconnTimes*2, self.__retry_connect_live_subscriber)
            reTrytimer.start()

    def __retry_connect_live_subscriber(self):
        """ 重连订阅服务 """

        if self.state == ERoomState.GameOver:
            return
        log.info("虎牙websocket断线重连 roomId:{}".format(self.roomId))
        self.__connect_live_subscriber()
        self.reconnTimes = self.reconnTimes + 1
        if self.reconnTimes >= 10:
            self.reconnTimes = 0
